### PR TITLE
Linting overhaul

### LINKED ISSUE #681: Use airbnb-eslint config
From https://github.com/exercism/javascript/pull/645

Turn on [airbnb-base](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb) for maintaining via a script that switches this.

In a later change, we should turn on several rules that guard against potential bugs, such as `no-shadow`, `no-undefined`, `no-var`; additionally we can add stylistic consistency without forcing style choises such as consistent spacing (but not required).

We should stay with `eslint:recommended` for exercises (end users) and use airbnb only for this repo. @SleeplessByte 

Wait for https://github.com/exercism/javascript/pull/680 to be merged so that we can exclude `airbnb` from `package.json` checksum CI task.

### PR DESCRIPTION
**Pull request by SleeplessByte**
REVIEWERS: you only need to review this message (the PR description). The PR itself is the implementation of this description + syncing files. If you really want to review it, you only need to look at [`89997af` (#1094)](https://github.com/exercism/javascript/pull/1094/commits/89997afef9d09132061f9d9881acbfdb9c8d0eca)

-------

This PR aims to further implement #44, improve on #645, and sorta fixes #681.

- Use a shared configuration for all the student editable files: `@exercism/eslint-config-javascript`
- Use a shared configuration for all read-only and maintainer only files: `@exercism/eslint-config-javascript/maintainers`
- Fix a bug where files inside `.meta` were not (always) linted
- Run the stricter configuration when doing `scripts/lint`

## Student rules

Furthermore, it enables the following rules (on top of the current recommended set) that often resolve actual errors (and are not really style dependent) for students:

### Warnings

- [`array-callback-return`](https://eslint.org/docs/rules/array-callback-return)
- [`eqeqeq`](https://eslint.org/docs/rules/eqeqeq)
- [`guard-for-in`](https://eslint.org/docs/rules/guard-for-in)
- [`no-extend-native`](https://eslint.org/docs/rules/no-extend-native)
- [`no-implicit-coercion`](https://eslint.org/docs/rules/no-implicit-coercion)
- [`no-promise-executor-return`](https://eslint.org/docs/rules/no-promise-executor-return)
- [`no-shadow`](https://eslint.org/docs/rules/no-shadow)
- [`no-unreachable-loop`](https://eslint.org/docs/rules/no-unreachable-loop)
- [`require-atomic-updates`](https://eslint.org/docs/rules/require-atomic-updates)

These will not block the student (ever), but almost *always* indicate mistakes. If you'd like to discuss why these were chosen, let me know, and I can give ample examples of actual student code as well as documented examples that these rules solve. Note that none of these rules are listed under "style" rules. In an upcoming update to the analyzer, these warnings will be shown to the user as `actionable`.

### Errors

- [`default-param-last`](https://eslint.org/docs/rules/default-param-last)
- [`no-eval`](https://eslint.org/docs/rules/no-eval)
- [`no-unsafe-optional-chaining`](https://eslint.org/docs/rules/no-unsafe-optional-chaining)

These are *always* mistakes, hence they are errors. In an upcoming update to the analyzer, these warnings will be shown to the user as `essential` and effectively soft-block the student.

### Maintainers / Contributors

Same list as for students but **everything** is an *error*. On top of that, the following rules are added to guard for mistakes with `import` and `export` statements, which are vital in the stubs and the test files.

In a previous track-wide update we've chosen for named exports (with one or more exceptions) which is why we discourage default exports. Additionally, the extension inclusion/exclusion is for consistency. The unresolved imports ensures all stubs export all the required bindings.

### Warnings

- [`import/no-default-export`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-default-export.md)

### Errors

- [`import/extensions`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md) with `{ js: 'never', json: 'always' }`
- [`import/no-unresolved`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unresolved)

Fixes #681

### CODE REVIEW

**Comment by SleeplessByte** [comment]
(I intend to merge this Friday the latest but feel free to voice your concerns, ideas, and/or anything else)

**Comment by neenjaw** [comment]
`reputation/contributed_code/double_major` <-- you should apply this label

**Comment by SleeplessByte** [comment]
@neenjaw does that exist?

**Comment by neenjaw** [comment]
@ErikSchierboom @iHiD
giphy (1)

**Review by ErikSchierboom — APPROVED** [review]

**Comment by junedev** [comment]
Great work! Thanks for taking the time to work through this.

(Possibly) Stupid question: Where is the source for @exercism/eslint-config-javascript? On npm the readme is empty and it does not show a repository.

I mostly like the rules you added.
I would not have picked guard-for-in though. Imo if someone just created a simple object literal and now wants to iterate over it, for-in without guard is just fine. This object literal scenario is also what is most likely to happen when a student uses for-in in an exercise. Forcing the guard although it is not necessary kind of defeats the point of for-in as a readable, short iteration helper. Imo for-in is demonized too much. Just like many things in JavaScript, it has its pitfalls/limitations.

The second thing I am not super happy about is the new formatting for the switch statements (not sure which rule is responsible for the change here). Tbh I never understood the advantage and I thought this "no indentation for the cases" style went out of fashion a while ago. Two reasons why I am not a fan. 1. In all other occurances we always indent the code inside a code block so it makes sense to be consistent. 2. Most docs, tutorials etc show the indented version. When a student comes from a language where indentations matters, the different versions might be confusing.

Feel free to ignore my comments and merge anyway. Just giving my 2 cents.

**Review by junedev — COMMENTED** [review]
Great work! Thanks for taking the time to work through this.

(Possibly) Stupid question: Where is the source for @exercism/eslint-config-javascript? On npm the readme is empty and it does not show a repository.

I mostly like the rules you added.
I would not have picked guard-for-in though. Imo if someone just created a simple object literal and now wants to iterate over it, for-in without guard is just fine. This object literal scenario is also what is most likely to happen when a student uses for-in in an exercise. Forcing the guard although it is not necessary kind of defeats the point of for-in as a readable, short iteration helper. Imo for-in is demonized too much. Just like many things in JavaScript, it has its pitfalls/limitations.

The second thing I am not super happy about is the new formatting for the switch statements (not sure which rule is responsible for the change here). Tbh I never understood the advantage and I thought this "no indentation for the cases" style went out of fashion a while ago. Two reasons why I am not a fan. 1. In all other occurances we always indent the code inside a code block so it makes sense to be consistent. 2. Most docs, tutorials etc show the indented version. When a student comes from a language where indentations matters, the different versions might be confusing.

Feel free to ignore my comments and merge anyway. Just giving my 2 cents.

**Comment by SleeplessByte** [comment]
> (Possibly) Stupid question: Where is the source for @exercism/eslint-config-javascript? On npm the readme is empty and it does not show a repository.

Waiting on Jeremy to create it for me.

- https://npm.runkit.com/@exercism/eslint-config-javascript/index.js?t=1618402770885
- https://npm.runkit.com/@exercism/eslint-config-javascript/maintainers.js?t=1618402770885

> I would not have picked guard-for-in though. Imo if someone just created a simple object literal and now wants to iterate over it, for-in without guard is just fine. This object literal scenario is also what is most likely to happen when a student uses for-in in an exercise. Forcing the guard although it is not necessary kind of defeats the point of for-in as a readable, short iteration helper. Imo for-in is demonized too much. Just like many things in JavaScript, it has its pitfalls/limitations.

@june Yeah. I struggled with this one. So I've seen mistakes in student code because they didn't know, but there are also so many cases where it's fine. I guess what we can do here is _remove_ this one from the config, but use the analyzer to show an informative pitfall if someone uses `for in`. Do you like that solution?

> The second thing I am not super happy about is the new formatting for the switch statements (not sure which rule is responsible for the change here). Tbh I never understood the advantage and I thought this "no indentation for the cases" style went out of fashion a while ago. Two reasons why I am not a fan. 1. In all other occurances we always indent the code inside a code block so it makes sense to be consistent. 2. Most docs, tutorials etc show the indented version. When a student comes from a language where indentations matters, the different versions might be confusing.

Nah, that's prettier, I think. I rather have it indent too. Can you point me to a line/file in this PR where this happens? I will try to find the tool responsible and "configure" it correctly, because I am 💯 % with you.

**Comment by junedev** [comment]
> I guess what we can do here is remove this one from the config, but use the analyzer to show an informative pitfall if someone uses for in.

Sounds like a good solution to me.

**Comment by SleeplessByte** [comment]
Weird. https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEAEnADYkQDqEATiQCYAUwA5tXHLlM-gLz4A6IABalyggDT5W7eHV74A5IKq06ghZIDOOEghjzUAQxKa4k6AFkI6UwHkAbnGriAvgEp8wflG-58mgHccGDAhBmkOHC4PLyg-PzBDU0URMggFJE9feL8Aelz-CABbSJ84+JdshKS4FJwMrPKc-MKS-AgYEWp8Tir8SvKBlxBxEAgAB1xoTWRQQ2pqCACABXmEGZRjAMMATxnRgCNqQzAAaw4AZXGTqOZkGGp0MxA4IoO4OjoPgBlDLnRDMw4AAxGhFQwwTh3TboGAQEbCGBFEgUITBOCaa5gOAXdbBHD2YI7ZDgTT7EBRUzUGDLY7McHIIwmZ4AK00AA8AELHM6XQwlb5ROCM4ymUZs9kXW66ACK6A6wqQTLFIGu1CpJIOhneJAR42oURgFBwdE6yAAHAAGUb6iCmCjHcYk-UYpyOBEAR3l8FpEw2ICSAFooOwvnQEWwvTg2LTAQylaLnqYijh7o8k9K4HKFSLmaMYNrjaahMgAEz5446W4AYWK8ZemgArAibHAACrajbK572J4ASSgX1gFzABsmAEFBxcYDtdLnTC4XEA seems correct to me. Is that not what you're expecting @junedev?

File: exercises/concept/mixed-juices/.meta/exemplar.js
Comment (junedev) [inline exercises/concept/mixed-juices/.meta/exemplar.js]: here is an example for the changed indentation

**Comment by junedev** [comment]
I looked at that commit that contains only the actual code changes that you posted. Maybe it was not reflecting the latest version. Playground example is correct.

**Comment by junedev** [comment]
I checked the latest version in the branch. Switch statement is correctly formatted there. Sorry for the confusion. 🙈

File: exercises/concept/mixed-juices/.meta/exemplar.js
Comment (SleeplessByte) [reply exercises/concept/mixed-juices/.meta/exemplar.js]: Thanks! Perfect. I will make sure this is right after we merge @junedev

**Comment by SleeplessByte** [comment]
> I checked the latest version in the branch. Switch statement is correctly formatted there. Sorry for the confusion. 🙈

I think you were right. So something removed the indentation and then formatting added it back 🤷🏽

**Comment by SleeplessByte** [comment]
Solved with https://github.com/exercism/eslint-config-javascript/pull/1

**Pull request merged**
